EventProc/src/OverlayProducer.cxx

```diff
@@ -0,0 +1,332 @@
+      overlayFileName_     = parameters.getParameter< std::string >("overlayFileName");
+      collections_         = parameters.getParameter< std::vector <std::string> >("overlayHitCollections");
+      simPassName_         = parameters.getParameter< std::string >( "passName" );
+      overlayPassName_     = parameters.getParameter< std::string >( "overlayPassName" );
+
+	  // overlay specifics: 
+      poissonMu_           = parameters.getParameter< double >("totalNumberOfInteractions");
+      doPoisson_           = parameters.getParameter< int >("doPoisson");
+      timeSigma_           = parameters.getParameter< double >("timeSpread");
+      timeMean_            = parameters.getParameter< double >("timeMean");
+	  nBunchesToSample_    = parameters.getParameter< int >("nBunchesToSample");
+	  bunchSpacing_        = parameters.getParameter< double >("bunchSpacing");
+	  
+      verbosity_           = parameters.getParameter< int >("verbosity");
+
+	  /// Print the parameters actually set. Helpful in case of typos.
+      if (verbosity_) {
+		 ldmx_log(info) <<  "Got parameters \n \t overlayFileName = " << overlayFileName_ 
+				  << "\n\t sim pass name = " << simPassName_
+				  << "\n\t overlay pass name = " << overlayPassName_
+				  << "\n\t overlayHitCollections = " ;
+		for ( const std::string &coll : collections_ )
+		   ldmx_log(info) <<  coll << "; " ;
+
+		ldmx_log(info) 
+				  << "\n\t numberOverlaidInteractions = " << poissonMu_
+				  << "\n\t doPoisson = " << doPoisson_
+				  << "\n\t timeSpread = " << timeSigma_
+				  << "\n\t timeMean = " << timeMean_
+				  << "\n\t verbosity = " << verbosity_ ;
+		
+      }
+	  
+	  return;
+    }
+  
+  void OverlayProducer::produce(Event& event) {
+	//event is the incoming, simulated event/"hard" process
+	//overlayEvent_ is the overlay producer's own event. 
+
+	if (verbosity_ > 1) {
+	   ldmx_log(info) <<  "produce() starts on simulation event " <<  event.getEventHeader().getEventNumber() ;
+	}
+
+	/// set up random seeds
+	if (rndm_.get() == nullptr) {
+	  //not been seeded yet, get it from RNSS
+	  const auto& rnss = getCondition<RandomNumberSeedService>(RandomNumberSeedService::CONDITIONS_OBJECT_NAME);
+	  rndm_ = std::make_unique<TRandom2>(rnss.getSeed("OverlayProducer::rndm"));
+	}
+	if (rndmTime_.get() == nullptr) {
+	  //not been seeded yet, get it from RNSS
+	  const auto& rnss = getCondition<RandomNumberSeedService>(RandomNumberSeedService::CONDITIONS_OBJECT_NAME);
+	  rndmTime_ = std::make_unique<TRandom2>(rnss.getSeed("OverlayProducer::rndmTime"));
+	}
+
+	
+	// sample a poisson distribution, or use a deterministic number of overlay events
+	int nEvsOverlay = doPoisson_ ? (int) rndm_->Poisson( poissonMu_ ) : (int)poissonMu_ ;
+	// the poisson samples the total number of events, which is nOverlay + 1 (since it includes the sim event)
+	nEvsOverlay -=1; // now subtract the sim event from the poisson mu
```

Comment:
What if there is no sim event before it? Could "inclusion" of the sim event be configurable?
Another option is to test if there is anything in the event bus using `exists`. For example, you could check for `SimParticles` which is a collection that is always produced by the simulation.

Reply:
So I've though about this, but rather from the angle that in a Poisson( mu < 4 or so), there will often be n_e = 0 events. But in MC, we always generate something that happens. So how to simulate this?

And why?
I think what we're possibly thinking about is an event where out-of-time (OOT) pileup signal triggers the event. An aside: I don't think we need to worry about simulating untriggered events -- we can get the total rate straight by weighting our events according to multiplicity, and accounting for that sometimes (a lot of the time) we will have 0e. And perhaps this is the way forward, just making fixed multiplicity samples and combining them, rather than Poisson fluctuating already at overlay stage (which mostly assures you that you don't waste time on multiplicities you don't care as much about). 

The case where OOT pileup triggers the event though. Wouldn't we have a possibility to study this basically in cases where the sim event itself wouldn't have triggered it, but (intime) pileup does, or when readout thresholds are already crossed by OOT pileup?

Or maybe we could speculate forever but just flying it once is the answer 😄  

I think it's an interesting point and possibly simple enough to implement but I'd like to have more of a direction decision for how to study OOT pileup before I go ahead and implement it. Maybe a separate issue for what you'd like to be able to study is a good idea, to make sure we don't lose track of the idea?

Reply:
I see your point. I will open another issue because it does get complicated rather quickly. :+1: